Keycloak self-managed documentation (#575)
elastisys/compliantkubernetes: docs/user-guide/additional-services/keycloak.md

> Keycloak (self-service)

Reviewer: Please add the following entry in `mkdocs.yml`, right below FerretDB:

```
      - 'Keycloak (self-service)': 'user-guide/additional-services/keycloak.md'
```

===========

> ![Keycloak Logo](img/keycloak.drawio.svg)

Reviewer: Please use this screenshot instead:
![screencapture-ck-keycloak-long-running-dev-ck8s-admin-master-console-2023-05-10-16_14_03](https://github.com/elastisys/compliantkubernetes/assets/1660833/cbbcb263-6bbf-4902-8df3-c79b7eec5ae2)

> ![Keycloak Image](img/keycloakimg.drawio.svg)

Reviewer: Any particular reason why the screenshot (PNG) is embedded in a DrawIO SVG? If not, please link to the screenshot PNG directly.

Author: Was just for consistency since all other images were that format. I will change to just linking the PNG directly.

This page will help you succeed in connecting your application to an identity and access management solution Keycloak, which meets your security and compliance requirements.

Keycloak is a widely recognized open-source Identity and Access Management (IAM) solution that provides user authentication and authorization services for applications.

> \

Reviewer: Please remove this line. It does not render well.
![image](https://github.com/elastisys/compliantkubernetes/assets/1660833/5cbd2d03-900c-428f-bcd5-20cb1fff48a4)

It offers a comprehensive set of features, including Single Sign-On (SSO), user federation, and social login support, making it a popular choice for securing applications in a variety of industries.

> \

Reviewer: Remove.

As of May 2023, Keycloak is a [CNCF Incubating project](https://www.cncf.io/blog/2023/04/11/keycloak-joins-cncf-as-an-incubating-project/).

In this guide we outline the necessary steps to configure and deploy a Keycloak instance on a Compliant Kubernetes cluster that is using the [managed PostgreSQL service](postgresql.md).

> \

Reviewer: Remove.

## Initial preparation
*Note: This guide assumes that you have managed PostgreSQL as an additional service.*

[Setup an application database and user in PostgreSQL](postgresql.md)

Take note of the following variables for the next section.

- The application secret you have created.

- The following environment variables:

```
echo $PGHOST
echo $APP_USERNAME
echo $APP_DATABASE
```

> 

Reviewer: Please keep only one empty line.

## Configure Keycloak with managed PostgreSQL

We chose Bitnami's Helm chart for [Keycloak](https://github.com/bitnami/charts/tree/main/bitnami/keycloak) due to its open-source nature, ease of deployment, security optimization, and active maintenance.

Bitnami is a well known provider of pre-configured, open-source application stacks that simplify deployment and management in various environments, such as Kubernetes. They offer a Helm chart for Keycloak, which streamlines deployment while adhering to Kubernetes best practices for security.

### Deploying Keycloak
Below is a sample **values.yaml** file that can be used to deploy Keycloak, please read the notes and change what is necessary.
```yaml
resources: # (1)
  limits:
    #cpu: 1000m
    memory: 1000Mi
  requests:
    cpu: 10m
    memory: 400Mi

postgresql:
  enabled: false

externalDatabase: # (2)
  host: "<PGHOST>"
  port: 5432
  user: <APP_USERNAME>
  database:  <APP_DATABASE>
  existingSecret: "<secret-name>"
  existingSecretPasswordKey: "PGPASSWORD"

ingress: # (3)
  enabled: true
  ingressClassName: "nginx"
  hostname: <URL>
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-prod
  tls: true

> 

Reviewer: Please keep only one empty line.

production: true # (4)
tls:
  enabled: true
  autoGenerated: true

> ```

Reviewer: This worked copy-paste "don't use brain". I love it!

However, the annotations are not rendered properly (see screenshot below). Can you check with `mkdocs serve`?

![image](https://github.com/elastisys/compliantkubernetes/assets/1660833/240b5562-3692-41a8-ba39-736c2902b829)

1.  The example provided serves as a starting point for configuring resource requests and limits for your Keycloak deployment. Be sure to tailor these values to your specific requirements, and monitor your deployment to optimize resource allocation for your unique use case.
2.  Insert the variables that you got from initial preparation.
3.  Configure the ingress hostname and which issuer you will be using.
4.  Enabling production mode and TLS for HTTPS. Disclaimer: Enabling production mode does not mean that the configuration here is ready for production. Please see further reading on Production configuration.

With the above values you can deploy Keycloak with:
```
helm repo add bitnami https://charts.bitnami.com/bitnami
helm upgrade --install keycloak bitnami/keycloak --values values.yaml
```

### Accessing Keycloak
When you have deployed Keycloak and can access the URL you can:

> \

Reviewer: Remove

[Login as admin and configure realms, users, and clients.](https://www.keycloak.org/getting-started/getting-started-kube)

> \

Reviewer: Remove.

> The default admin user is “user”.

Reviewer: ```suggestion
The default admin user is `user`.
```

The initial admin password can be fetched using the command below (assuming default namespace):

> ```

Reviewer: ```suggestion
```console
```

> password=$(kubectl get secret keycloak -o jsonpath="{.data.admin-password}" | base64 -d)

Reviewer: ```suggestion
kubectl get secret keycloak -o jsonpath="{.data.admin-password}" | base64 -d
```

> echo “Password: $password”

Reviewer: ```suggestion
```

Keycloak (self-managed)
===========

> # (Picture goes here)

Reviewer: Yes, thank you! Preferably include a screenshot you produced yourself after you ran the steps below.

This page will help you succeed in connecting your application to an identity and access management solution Keycloak, which meets your security and compliance requirements.

Keycloak is a widely recognized open-source Identity and Access Management (IAM) solution that provides user authentication and authorization services for applications. It offers a comprehensive set of features, including Single Sign-On (SSO), user federation, and social login support, making it a popular choice for securing applications in a variety of industries. In this guide we outline the necessary steps to configure and deploy a Keycloak instance on a Compliant Kubernetes cluster that is using the [managed PostgreSQL.](https://elastisys.io/compliantkubernetes/user-guide/additional-services/postgresql/) This will provide you with a robust and secure IAM solution to manage user access and authorization for your applications running on Compliant Kubernetes.
## Initial preparation
*Note: This guide assumes that you have managed PostgreSQL as an additional service.*

> 1. [Create a HNC namespace for Keycloak.](https://elastisys.io/compliantkubernetes/user-guide/namespaces/)

Reviewer: Let's skip this step. I see no problem with running Keycloak in my current namespace.

> 2. [Setup an application database and user in postgreSQL](https://elastisys.io/compliantkubernetes/user-guide/additional-services/postgresql)

Reviewer: Please be mindful about capitalization of brands (feels like I should add a pre-commit hook). :smile:

Also, for internal links, make sure to refer to the file, so that mkdocs can make internal link consistency checks.

```suggestion
2. [Setup an application database and user in PostgreSQL](postgresql.md)
```

> 3. Take note of the following variables from step 2 for the next section.

Reviewer: Can you produce a one-liner (perhaps broken across multiple lines for readability) like for FerretDB: https://elastisys.io/compliantkubernetes/user-guide/additional-services/ferretdb/#install

- The host of the postgres (PGHOST)
- The user you have created(APP_USER)
- The database you have created(APP_DATABASE)
- The application secret you have created

## Configure Keycloak with managed Postgres

We chose Bitnami's Helm chart for [Keycloak](https://github.com/bitnami/charts/tree/main/bitnami/keycloak) due to its open-source nature, ease of deployment, security optimization, and active maintenance. 
Bitnami is a well known provider of pre-configured, open-source application stacks that simplify deployment and management in various environments, such as Kubernetes. They offer a Helm chart for Keycloak, which streamlines deployment while adhering to Kubernetes best practices for security.

> 1. Add resource requests(and optionally limits) to the helm chart values.

Reviewer: For both step 1 and 2, can you produce a good starting `values.yaml` file which I can just copy-paste-adjust?

Please annotate the values file, as described [here](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations) and exemplified [here](https://elastisys.io/compliantkubernetes/user-guide/network-model/#issuing-a-certificate).

Then include a line like `helm upgrade --install ... -f values.yaml`.

```
resources:
  limits:
    #cpu: 1000m
    memory: 1000Mi
  requests:
    cpu: 10m
    memory: 400Mi
```
The example provided above serves as a starting point for configuring resource requests and limits for your Keycloak deployment. Be sure to tailor these values to your specific requirements, and monitor your deployment to optimize resource allocation for your unique use case.

2. Configure Keycloak values to use an external database based on the variables noted in the previous step.
```
postgresql:
 enabled: false

externalDatabase:
 host: "<PGHOST>"
 port: 5432
 user: <APP_USER>
 database: <APP_DATABASE>
 existingSecret: "<application-secret>"
 existingSecretPasswordKey: "PGPASSWORD"
```

3. Configure your values for production
\
a. https://www.keycloak.org/server/configuration-production  
b. Enable and configure an ingress.

> 4. Consider how you will be using Keycloak.

Reviewer: Please enable the Ingress of the Helm Chart. Make sure it has the right annotations for HTTPS:

```
    cert-manager.io/cluster-issuer: letsencrypt-prod
    kubernetes.io/ingress.class: nginx
```

\
a. [With a reverse proxy?](https://www.keycloak.org/server/configuration-production#_reverse_proxy_in_a_distributed_environment)
\
b. [By securing your applications?](https://www.keycloak.org/docs/latest/securing_apps/index.html)

5. When you have deployed Keycloak you can:
\
[Login as admin and configure realms, users, and clients.](https://www.keycloak.org/getting-started/getting-started-kube)
\

> The initial admin user is “user” and the initial admin password will be generated in the secret “keycloak”. This can be customized in the values if you want to generate your own admin password or change admin username.

Reviewer: Can you provide a one-liner to retrieve said password? Alternatively, the Helm Chart might already output that command. Can you then just copy-paste that output and refer to it?

> ## Demo?

Reviewer: Yes, please demo this during our next end-of-sprint demo.

Please include a "Further Reading" section here.
